packages/core/scripts/liquidity-mining/CalculateRollBalancerLPRewards.js

```diff
@@ -0,0 +1,250 @@
+
+  const shareHolderPayout = await _calculatePayoutsBetweenBlocks(
+    bPool1,
+    bPool2,
+    shareHolders,
+    fromBlock,
+    toBlock,
+    blocksPerSnapshot,
+    umaPerSnapshot,
+    snapshotsToTake
+  );
+
+  console.log("🎉 Finished calculating payouts!");
+  _saveShareHolderPayout(
+    shareHolderPayout,
+    rollNum,
+    fromBlock,
+    toBlock,
+    tokenName,
+    pool1Address,
+    pool2Address,
+    umaPerWeek,
+    blocksPerSnapshot
+  );
+}
+
+// Calculate the payout to a list of `shareHolders` between `fromBlock` and `toBlock`. Split the block window up into
+// chunks of `blockPerSnapshot` and at each chunk assign `umaPerSnapshot` at a prorata basis.
+async function _calculatePayoutsBetweenBlocks(
+  bPool1,
+  bPool2,
+  shareHolders,
+  fromBlock,
+  toBlock,
+  blockPerSnapshot,
+  umaPerSnapshot
+) {
+  // Create a structure to store the payouts for all historic shareholders.
+  let shareHolderPayout = {};
+  for (shareHolder of shareHolders) {
+    shareHolderPayout[shareHolder] = toBN("0");
+  }
+
+  console.log("🏃‍♂️Iterating over block range and calculating payouts...");
+
+  // create new progress bar to show the status of blocks traversed.
+  const progressBar = new cliProgress.SingleBar(
+    {
+      format: "[{bar}] {percentage}% | snapshots traversed: {value}/{total}"
+    },
+    cliProgress.Presets.shades_classic
+  );
+  progressBar.start(Math.ceil((toBlock - fromBlock) / blockPerSnapshot), 0);
+  for (currentBlock = fromBlock; currentBlock < toBlock; currentBlock += blockPerSnapshot) {
+    shareHolderPayout = await _updatePayoutAtBlock(bPool1, bPool2, currentBlock, shareHolderPayout, umaPerSnapshot);
+    progressBar.update(Math.ceil((currentBlock - fromBlock) / blockPerSnapshot) + 1);
+  }
+  progressBar.stop();
+
+  return shareHolderPayout;
+}
+
+// For a given `blockNumber` (snapshot in time), return an updated `shareHolderPayout` object that has appended
+// payouts for a given `bPool` at a rate of `umaPerSnapshot`.
+async function _updatePayoutAtBlock(bPool1, bPool2, blockNumber, shareHolderPayout, umaPerSnapshot) {
+  // Get the total supply of Balancer Pool tokens at the given snapshot's block number.
+  const bptSupplyAtSnapshot = toBN(await bPool1.methods.totalSupply().call(undefined, blockNumber)).add(
+    toBN(await bPool2.methods.totalSupply().call(undefined, blockNumber))
+  );
+
+  // Get the given holders balance at the given block. Generate an array of promises to resolve in parallel.
+  let promiseArraybPool1 = [];
+  let promiseArraybPool2 = [];
+  for (shareHolder of Object.keys(shareHolderPayout)) {
+    promiseArraybPool1.push(bPool1.methods.balanceOf(shareHolder).call(undefined, blockNumber));
+    promiseArraybPool2.push(bPool2.methods.balanceOf(shareHolder).call(undefined, blockNumber));
+  }
+  const balanceResultsbPool1 = await Promise.allSettled(promiseArraybPool1);
+  await delay(5); // slow down the queries between these massive promise arrays. Helps to keep Infura happy.
```

Comment:
nit: what does infura do if you don't?

Would bluebird promises be a better way to throttle the total number of outstanding requests?

Reply:
bump.

Reply:
I implemented bluebird promises but I found that this timed out more and was slower. For some reason, infura is ok with a large burst (~2000 calls in parallel) if you then back off for a bit.